### PR TITLE
pgremapper drain: support draining from multiple sources

### PR DESCRIPTION
**Pull request by DanWritesCode**
Allows you to specify multiple OSDs, or an entire osdspec as your source for the 'drain' sub-command.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -152,41 +152,45 @@ scheduled backfills. No attempt is made to balance the fullness of the target
...
 			}
 
+			for _, osd := range sourceOsds {
+				sourceOsdNode, ok := tree.IDToNode[osd]
+				if !ok || sourceOsdNode.Type != "osd" {
+					panic(fmt.Errorf("source OSD %d doesn't exist", osd))
+				}
+				delete(targetOsds, osd)
+			}
+
```
Comment (baergj) [inline main.go:190]: Make sure to document this behaviour and other changes in the command help and README.

File: README.md [hunk 2]
```diff
@@ -146,10 +146,11 @@ $ ./pgremapper cancel-backfill --pgs-including bucket:data10
 
 ### drain
 
-Remap PGs off of the given source OSD, up to the given maximum number of scheduled backfills. No attempt is made to balance the fullness of the target OSDs; rather, the least busy target OSDs and PGs will be selected.
+Remap PGs off of the given source OSD spec(s), up to the given maximum number of scheduled backfills. No attempt is made to balance the fullness of the target OSDs; rather, the least busy target OSDs and PGs will be selected.
+If a source OSD is included among target OSDs, it will be removed from the targets.
 
 ```
-$ ./pgremapper drain <source OSD> --target-osds <osdspec>[,<osdspec>] [--allow-movement-across <bucket type>] [--max-backfill-reservations default_max[,osdspec:max]] [--max-source-backfills <n>]
+$ ./pgremapper drain <osdspec>[,<osdspec>] --target-osds <osdspec>[,<osdspec>] [--allow-movement-across <bucket type>] [--max-backfill-reservations default_max[,osdspec:max]] [--max-source-backfills <n>]
```
Comment (baergj) [inline README.md:153]: I think the osdspecs are actually space-separated, no? And for consistency with other commands I think this should be `<osdspec> ...`.

Comment (DanWritesCode) [reply README.md:153]: oh, I copied that `[,osdspec]` part from the undo-upmaps documentation in the readme... let me get that fixed up

File: README.md [hunk 3]
```diff
@@ -150,7 +150,7 @@ Remap PGs off of the given source OSD spec(s), up to the given maximum number of
 If a source OSD is included among target OSDs, it will be removed from the targets.
 
 ```
-$ ./pgremapper drain <osdspec>[,<osdspec>] --target-osds <osdspec>[,<osdspec>] [--allow-movement-across <bucket type>] [--max-backfill-reservations default_max[,osdspec:max]] [--max-source-backfills <n>]
+$ ./pgremapper drain <osdspec> [<osdspec> ...] --target-osds <osdspec>[,<osdspec>] [--allow-movement-across <bucket type>] [--max-backfill-reservations default_max[,osdspec:max]] [--max-source-backfills <n>]
```
Comment (baergj) [inline README.md:153]: This is different from how we wrote this for `export-mappings`, at least, so a general harmonization would be appreciated.

File: main.go [hunk 4]
```diff
@@ -143,7 +143,7 @@ has been made so far.
 	}
 
 	drainCmd = &cobra.Command{
-		Use:   "drain <source osdspec>",
+		Use:   "drain [<osdspec> ...]",
```
Comment (baergj) [inline main.go:146]: This is different from the README

File: main.go [hunk 5]
```diff
@@ -152,41 +152,45 @@ scheduled backfills. No attempt is made to balance the fullness of the target
 OSDs; rather, the least busy target OSDs and PGs will be selected.
 `,
 		Args: func(cmd *cobra.Command, args []string) error {
-			if len(args) != 1 {
+			if len(args) == 0 {
 				return errors.New("a source OSD must be specified")
```
[the remark this answers is not in the document]
Comment (baergj) [reply main.go:156]: I think this is still outstanding

File: README.md (hunk 3 above)
Comment (DanWritesCode) [reply README.md:153]: looks like a couple things slipped through

File: main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanWritesCode) [reply main.go]: ah, yes, I re-worded it above but didnt re-word the error itself

File: README.md [hunk 6]
```diff
@@ -179,10 +180,10 @@ Export all upmaps for the given OSD spec(s) in a json format usable by import-ma
 Note that the mappings exported will be just the portions of the upmap items pertaining to the selected OSDs (i.e. if a given OSD is the From or To of the mapping), unless `--whole-pg` is specified.
 
 ```
-$ ./pgremapper export-mappings <osdspec> ... [--output <file>] [--whole-pg]
+$ ./pgremapper export-mappings <osdspec> [<osdspec> ...] [--output <file>] [--whole-pg]
```
Comment (baergj) [inline README.md:183]: Does the code docs need an update here too?

I'll stop being a pain now :)

**Review by baergj — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (DanWritesCode) [reply README.md:183]: haha no worries
https://github.com/digitalocean/pgremapper/blob/7928d2f918ff9de9a79b7c0335113de5ac28c9f2/main.go#L306-L310
It LGTM

**Pull request merged**